torch/utils/_pytree.py

```diff
@@ -452,52 +452,86 @@ def _dict_unflatten(values: Iterable[Any], context: Context) -> Dict[Any, Any]:
 
 
 def _namedtuple_flatten(d: NamedTuple) -> Tuple[List[Any], Context]:
-    return list(d), type(d)
+    return list(d), {"type": type(d), "fields": list(d._fields)}
```

Comment:
> We want to update pytree TreeSpec's context for NamedTuple to include the field names for better BC checking. If a user in an older runtime uses a NamedTuple with 2 fields ["a", "b"] at one point in time, serializes this spec, and then loads it up into a newer runtime where the NamedTuple now has fields ["a", "c", "b"], we want to be able to show that "c" is the field which has been newly added.

To achieve this, we only need to change the serialization logic for namedtuple context while keeping the context unchanged (i.e., the namedtuple type). Because the type class contains all the information you need where `fields = type._fields`.

Reply:
Serialization/deserialization doesn't provide any entrypoints for users to inspect the differing specs. Given an input in the new runtime, they want to adapt the input to the original spec, so having the field information then is very useful.

Reply:
You only need to update `_namedtuple_serialize` and `_namedtuple_deserialize`.

- `_namedtuple_serialize`: dump `context._fields` to json spec.
- `_namedtuple_deserialize`: emit a warning if the number of fields in the json spec does not match the length of namedtuple type.

Reply:
> _namedtuple_deserialize: emit a warning if the number of fields in the json spec does not match the length of namedtuple type.

But where will we store the field names in the deserialized treespec? We need to be able to compare treespecs and get the concrete field names which are different so that we can adapt the inputs to fit the old treespec.

Reply:
> But where will we store the field names in the deserialized treespec? We need to be able to compare treespecs and get the concrete field names which are different so that we can adapt the inputs to fit the old treespec.

Do you mean the reconstructed treespec will ignore the newly added fields?

```python
def _namedtuple_unflatten(values: Iterable[Any], context: Context) -> NamedTuple:
   return context["type"](**dict(zip(context["fields"], values)))
```

Reply:
> Hmm it doesn't seem right that the field names show up in the context only when we serialize.

If you just want to prompt the user the fields of the namedtuple types, you can use `treespec_pprint(...)`.

```python
In [1]: from collections import namedtuple

In [2]: MyTuple = namedtuple('MyTuple', ['a', 'b', 'c'])

In [3]: t = MyTuple(1, 2, [3, 4])

In [4]: import torch.utils._pytree as pytree

In [5]: spec = pytree.tree_structure(t)

In [6]: spec
Out[6]: 

TreeSpec(namedtuple, <class '__main__.MyTuple'>, [*,
  *,
  TreeSpec(list, None, [*,
│   *])])

In [7]: pytree.treespec_pprint(spec)
Out[7]: 'MyTuple(a=*, b=*, c=[*, *])'
```

Reply:
No, we want to get the field names from the old runtime. This will give the field names of the current runtime?

Reply:
No, we want to get the field names from the old runtime. This will give the field names of the current runtime?